WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt

```diff
@@ -0,0 +1,159 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
+
+/**
+ * The logic for the speaker sample.
+ *
+ * The stateful logic is kept by a [MainStateHolder].
+ */
+@Composable
+fun SpeakerApp() {
+    MaterialTheme {
+        lateinit var requestPermissionLauncher: ManagedActivityResultLauncher<String, Boolean>
+
+        val context = LocalContext.current
+        val activity = context.findActivity()
+        val coroutineScope = rememberCoroutineScope()
+
+        val stateHolder = remember(activity) {
+            MainStateHolder(
+                activity = activity,
+                requestPermission = {
+                    requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                },
+                showPermissionRationale = {
+                    // TODO: Refactor away from normal AlertDialog to a Compose-specific dialog
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.rationale_for_microphone_permission)
+                        .setPositiveButton(R.string.ok) { _, _ ->
+                            requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                        }
+                        .setNegativeButton(R.string.cancel) { _, _ -> }
+                        .create()
+                        .show()
+                },
+                showSpeakerNotSupported = {
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.no_speaker_supported)
+                        .setPositiveButton(R.string.ok) { _, _ -> }
+                        .create()
+                        .show()
+                }
+            )
+        }
+
+        val lifecycleOwner = LocalLifecycleOwner.current
+
+        requestPermissionLauncher = rememberLauncherForActivityResult(RequestPermission()) {
+            // We ignore the direct result here, since we're going to check anyway.
+            coroutineScope.launch {
+                val job = launch { stateHolder.onAction(AppAction.PermissionResultReturned) }
+                lifecycleOwner.lifecycle.awaitEvent(Lifecycle.Event.ON_STOP)
+                job.cancel()
+            }
+        }
+
+        // Notify the state holder whenever we become started to reset the state
+        LaunchedEffect(stateHolder) {
+            lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                stateHolder.onAction(AppAction.Started)
+            }
+        }
+
+        SpeakerScreen(
+            appState = stateHolder.appState,
+            isPermissionDenied = stateHolder.isPermissionDenied,
+            recordingProgress = stateHolder.recordingProgress,
+            onMicClicked = {
+                coroutineScope.launch {
+                    val job = launch { stateHolder.onAction(AppAction.MicClicked) }
+                    lifecycleOwner.lifecycle.awaitEvent(Lifecycle.Event.ON_STOP)
```

Comment:
I'm curious about this pattern. onAction doesn't look like a suspending function, does the cancellation interrupt the onAction method?

Also wondering why launching with coroutineScope instead of either LaunchedEffect or using LocalLifecycleOwner.current.lifecycleScope which I assume would handle the cancellation?

I'm relatively new to Compose, so mainly trying to understand why the simpler forms don't work here.  Thanks for your patience :)

Reply:
> I'm relatively new to Compose
So am I, and aren't we all 😄 

`onAction` is a suspending function, which will modify the underlying state, as well as asynchronously do things in response to the user action. In particular, this allows tying the cancellation of `onAction` to cancelling recording and playback, since those are encapsulated as suspending functions as well.

So in this case (assuming the user has granted permission), `onAction(MicClicked)` will kick off the recording, updating the overall state to be `recording`, updating the progress for the duration of the recording, and resetting the layout when recording is finished.

Since we're doing all of that in response to a user action, I'm launching into a `rememberCoroutineScope()` (as opposed to a `LaunchedEffect` or similar, which I would do if it was in response to something other than a user action).

There are two other behaviors that also complicate this:
- Inside the state holder is a [`MutatorMutex`](https://developer.android.com/reference/kotlin/androidx/compose/foundation/MutatorMutex), provided by Compose. This allows cancelling other coroutines, so we get a lot of the state management for free: If the user clicks record, and then cancels it, the `onAction` for recording will be automatically cancelled, and cleaned up.
- I want recording and playback to be cancelled upon the app going to the background (or in other words, I only want it to occur when at least started). So the `lifecycleScope` would allow for nicely cleaning everything up once destroyed, but not as nicely for stopped.

Reply:
Thinking about this more, there's one thing my current approach doesn't do:

If the `onAction` completes successfully, we don't really have to keep waiting for the `ON_STOP` even. Really what we want to do is ["race"](https://github.com/Kotlin/kotlinx.coroutines/issues/2867) waiting for `ON_STOP` and the `onAction` call.

Whichever happens first, we want to cancel the other.

Reply:
What would break if you changed

```
            onMicClicked = {
                scope.launch {
                    raceOf(
                        { stateHolder.onAction(AppAction.MicClicked) },
                        { lifecycleOwner.lifecycle.awaitEvent(Lifecycle.Event.ON_STOP) }
                    )
                }
            },
```

to

```
            onMicClicked = {
                scope.launch {
                     stateHolder.onAction(AppAction.MicClicked)
                }
            },
```            

Does rememberCoroutineScope get closed when the overall Activity is stopped, or the @Composable is removed form the UI?

This is the sort of nuanced code that seems difficult to get right. As a sample I hope (but don't know if possible) that there is some simple obvious compose idiomatic way where this just works without everyone writing similarly complex code.

Reply:
I wonder whether lifecycle-runtime-ktx is missing something here.  It feels like what you are trying is directly in scope.

https://developer.android.com/topic/libraries/architecture/lifecycle
> Pausing and resuming animated drawables. Use lifecycle-aware components to handle pausing animated drawables when while app is in the background and resume drawables after the app is in the foreground.

Reply:
I wonder whether `withStateAtLeast` could be the solution here? As a sample, I'm concerned if you need to understand this complicated state dance, others will to.

Reply:
Yeah, `withStateAtLeast` almost does what I want, but the `block` passed to it can't be suspending.

If there was a version of `withStateAtLeast` that did take a suspending block, that'd roughly be the behavior I'd expect:
wait for the state to be met, start the block, and cancel it if dropping below the given state.

Reply:
Thanks for humoring me with these questions...